File: packages/domain/src/sales/sale-item.ts
import type { ProductId } from '../value-objects/ids.js';
import { Money } from '../value-objects/money.js';
import { Quantity } from '../value-objects/quantity.js';

/** Single product line on a Sale — quantity, unit price, line subtotal. */
export class SaleItem {
  private constructor(
    readonly productId: ProductId,
    readonly quantity: Quantity,
    readonly unitPrice: Money,
    readonly lineTotal: Money,
  ) {}

  static create(productId: ProductId, quantity: Quantity, unitPrice: Money): SaleItem {
    const lineTotal = unitPrice.multiply(quantity.value);
    return new SaleItem(productId, quantity, unitPrice, lineTotal);
  }
}

File: backend/crates/domain-sales/src/sale_item.rs
use domain_inventory::{ProductId, Quantity};
use domain_shared::Money;

/// Single product line on a Sale — quantity, unit price, line subtotal.
#[derive(Debug, Clone)]
pub struct SaleItem {
    product_id: ProductId,
    quantity: Quantity,
    unit_price: Money,
    line_total: Money,
}

impl SaleItem {
    pub fn create(
        product_id: ProductId,
        quantity: Quantity,
        unit_price: Money,
    ) -> Result<Self, domain_shared::DomainError> {
        let line_total = unit_price
            .amount_minor()
            .checked_mul(quantity.value() as i64)
            .and_then(|amount| Money::new(amount, unit_price.currency()).ok())
            .ok_or(domain_shared::DomainError::MoneyOverflow)?;
        Ok(Self {
            product_id,
            quantity,
            unit_price,
            line_total,
        })
    }

    pub fn product_id(&self) -> ProductId {
        self.product_id
    }

    pub fn quantity(&self) -> Quantity {
        self.quantity
    }

    pub fn unit_price(&self) -> &Money {
        &self.unit_price
    }

    pub fn line_total(&self) -> &Money {
        &self.line_total
    }
}
